File: recapit/quota.py
from __future__ import annotations

import logging
import time
from collections import deque
from contextlib import contextmanager
from dataclasses import dataclass
from threading import Lock
from typing import Deque, Dict, Iterable, Mapping

logger = logging.getLogger(__name__)


@dataclass
class QuotaConfig:
    request_limits: Mapping[str, int]
    token_limits: Mapping[str, int]
    rpm_warn_threshold: float = 0.8
    rpm_sleep_threshold: float = 0.9
    token_warn_threshold: float = 0.8
    storage_limit_bytes: int = 20 * 1024 * 1024 * 1024  # 20 GB
    upload_limit_bytes: int = 2 * 1024 * 1024 * 1024  # 2 GB per upload
    concurrency_limit: int = 100
    warn_cooldown_seconds: float = 10.0
    max_preemptive_sleep: float = 0.5


class QuotaMonitor:
    def __init__(self, config: QuotaConfig) -> None:
        self._config = config
        self._lock = Lock()
        self._token_windows: Dict[str, Deque[tuple[float, int]]] = {}
        self._token_warned_at: Dict[str, float] = {}
        self._rpm_warned_at: Dict[str, float] = {}
        self._uploaded_bytes: int = 0
        self._active_uploads: int = 0

    def check_rpm(self, model: str, utilization: float, per_minute: int, window_seconds: int) -> None:
        now = time.monotonic()
        sleep_duration = 0.0
        with self._lock:
            if utilization >= self._config.rpm_warn_threshold:
                last_warn = self._rpm_warned_at.get(model, 0.0)
                if now - last_warn >= self._config.warn_cooldown_seconds:
                    logger.warning(
                        "Model %s request rate at %.0f%% of per-minute quota (%d RPM)",
                        model,
                        utilization * 100.0,
                        per_minute,
                    )
                    self._rpm_warned_at[model] = now
            if utilization >= self._config.rpm_sleep_threshold:
                sleep_duration = min(window_seconds / max(per_minute, 1), self._config.max_preemptive_sleep)
        if sleep_duration > 0:
            time.sleep(sleep_duration)

    def register_event(self, *, model: str, timestamp: float, total_tokens: int | None) -> None:
        if total_tokens is None:
            return
        limit = self._config.token_limits.get(model)
        if not limit:
            return
        with self._lock:
            window = self._token_windows.setdefault(model, deque())
            window.append((timestamp, total_tokens))
            cutoff = timestamp - 60.0
            while window and window[0][0] < cutoff:
                window.popleft()
            window_total = sum(tokens for _, tokens in window)
            utilization = window_total / limit
            if utilization >= self._config.token_warn_threshold:
                last_warn = self._token_warned_at.get(model, 0.0)
                if timestamp - last_warn >= self._config.warn_cooldown_seconds:
                    logger.warning(
                        "Model %s token usage at %.0f%% of per-minute quota (%d tokens/min)",
                        model,
                        utilization * 100.0,
                        limit,
                    )
                    self._token_warned_at[model] = timestamp

    @contextmanager
    def track_upload(self, *, path: str, size_bytes: int) -> Iterable[None]:
        self._before_upload(path=path, size_bytes=size_bytes)
        try:
            yield
        finally:
            self._after_upload(size_bytes=size_bytes)

    def _before_upload(self, *, path: str, size_bytes: int) -> None:
        if size_bytes > self._config.upload_limit_bytes:
            raise ValueError(
                f"Upload {path} exceeds per-file upload limit of {self._config.upload_limit_bytes // (1024 * 1024 * 1024)} GB"
            )
        with self._lock:
            self._uploaded_bytes += size_bytes
            self._active_uploads += 1
            storage_util = self._uploaded_bytes / self._config.storage_limit_bytes
            if storage_util >= self._config.token_warn_threshold:
                logger.warning(
                    "Uploads this run total %.2f GB (%.0f%% of 20 GB Files API window)",
                    self._uploaded_bytes / (1024**3),
                    storage_util * 100.0,
                )
            concurrency_util = self._active_uploads / max(self._config.concurrency_limit, 1)
            if concurrency_util >= self._config.token_warn_threshold:
                logger.warning(
                    "Concurrent uploads at %d/%d (%.0f%% of limit)",
                    self._active_uploads,
                    self._config.concurrency_limit,
                    concurrency_util * 100.0,
                )

    def _after_upload(self, *, size_bytes: int) -> None:
        with self._lock:
            self._active_uploads = max(self._active_uploads - 1, 0)

File: src/quota.rs
use std::collections::{HashMap, VecDeque};
use std::sync::{Arc, Mutex};
use std::time::{Duration, Instant};

use anyhow::{bail, Result};
use tracing::warn;

#[derive(Debug, Clone)]
pub struct QuotaConfig {
    pub request_limits: HashMap<String, u32>,
    pub token_limits: HashMap<String, u32>,
    pub rpm_warn_threshold: f64,
    pub rpm_sleep_threshold: f64,
    pub token_warn_threshold: f64,
    pub storage_limit_bytes: u64,
    pub upload_limit_bytes: u64,
    pub concurrency_limit: u32,
    pub warn_cooldown: Duration,
    pub max_preemptive_sleep: Duration,
    pub request_window: Duration,
}

impl QuotaConfig {
    pub fn new(request_limits: HashMap<String, u32>, token_limits: HashMap<String, u32>) -> Self {
        Self {
            request_limits,
            token_limits,
            rpm_warn_threshold: 0.8,
            rpm_sleep_threshold: 0.9,
            token_warn_threshold: 0.8,
            storage_limit_bytes: 20 * 1024 * 1024 * 1024,
            upload_limit_bytes: 2 * 1024 * 1024 * 1024,
            concurrency_limit: 100,
            warn_cooldown: Duration::from_secs(10),
            max_preemptive_sleep: Duration::from_millis(500),
            request_window: Duration::from_secs(60),
        }
    }
}

#[derive(Default)]
struct QuotaState {
    request_windows: HashMap<String, VecDeque<Instant>>,
    last_rpm_warn: HashMap<String, Instant>,
    token_windows: HashMap<String, VecDeque<(Instant, u32)>>,
    last_token_warn: HashMap<String, Instant>,
    uploaded_bytes: u64,
    active_uploads: u32,
}

#[derive(Clone)]
pub struct QuotaMonitor {
    config: Arc<QuotaConfig>,
    state: Arc<Mutex<QuotaState>>,
}

impl QuotaMonitor {
    pub fn new(config: QuotaConfig) -> Self {
        Self {
            config: Arc::new(config),
            state: Arc::new(Mutex::new(QuotaState::default())),
        }
    }

    pub fn register_request(&self, model: &str) -> Option<Duration> {
        let per_minute = match self.config.request_limits.get(model) {
            Some(value) if *value > 0 => *value,
            _ => return None,
        };
        let mut state = self.state.lock().unwrap();
        let window = state.request_windows.entry(model.to_string()).or_default();
        let now = Instant::now();
        window.push_back(now);
        while let Some(front) = window.front() {
            if now.duration_since(*front) > self.config.request_window {
                window.pop_front();
            } else {
                break;
            }
        }
        let utilization = window.len() as f64 / per_minute as f64;
        if utilization >= self.config.rpm_warn_threshold {
            let entry = state
                .last_rpm_warn
                .entry(model.to_string())
                .or_insert(now - self.config.warn_cooldown * 2);
            if now.duration_since(*entry) >= self.config.warn_cooldown {
                warn!(
                    "model {} request rate at {:.0}% of limit ({} RPM)",
                    model,
                    utilization * 100.0,
                    per_minute
                );
                *entry = now;
            }
        }
        if utilization >= self.config.rpm_sleep_threshold {
            let window_seconds = self.config.request_window.as_secs_f64();
            let per_request = if per_minute > 0 {
                window_seconds / per_minute as f64
            } else {
                0.0
            };
            let sleep = self
                .config
                .max_preemptive_sleep
                .as_secs_f64()
                .min(per_request);
            if sleep > 0.0 {
                return Some(Duration::from_secs_f64(sleep));
            }
        }
        None
    }

    pub fn register_tokens(&self, model: &str, total_tokens: Option<u32>) {
        let Some(total_tokens) = total_tokens else {
            return;
        };
        let limit = match self.config.token_limits.get(model) {
            Some(value) if *value > 0 => *value,
            _ => return,
        };
        let mut state = self.state.lock().unwrap();
        let window = state.token_windows.entry(model.to_string()).or_default();
        let now = Instant::now();
        window.push_back((now, total_tokens));
        while let Some((instant, _)) = window.front() {
            if now.duration_since(*instant) > self.config.request_window {
                window.pop_front();
            } else {
                break;
            }
        }
        let used: u64 = window.iter().map(|(_, tokens)| *tokens as u64).sum();
        let utilization = used as f64 / limit as f64;
        if utilization >= self.config.token_warn_threshold {
            let entry = state
                .last_token_warn
                .entry(model.to_string())
                .or_insert(now - self.config.warn_cooldown * 2);
            if now.duration_since(*entry) >= self.config.warn_cooldown {
                warn!(
                    "model {} token usage at {:.0}% of per-minute quota ({} tokens/min)",
                    model,
                    utilization * 100.0,
                    limit
                );
                *entry = now;
            }
        }
    }

    pub fn track_upload(&self, path: &str, size_bytes: u64) -> Result<UploadGuard> {
        if size_bytes > self.config.upload_limit_bytes {
            bail!(
                "Upload {} exceeds per-file upload limit of {} GB",
                path,
                self.config.upload_limit_bytes as f64 / (1024.0 * 1024.0 * 1024.0)
            );
        }
        let mut state = self.state.lock().unwrap();
        state.uploaded_bytes += size_bytes;
        state.active_uploads += 1;

        let storage_util = state.uploaded_bytes as f64 / self.config.storage_limit_bytes as f64;
        if storage_util >= self.config.token_warn_threshold {
            warn!(
                "Uploads this run total {:.2} GB ({:.0}% of {:.0} GB Files API window)",
                state.uploaded_bytes as f64 / (1024.0 * 1024.0 * 1024.0),
                storage_util * 100.0,
                self.config.storage_limit_bytes as f64 / (1024.0 * 1024.0 * 1024.0)
            );
        }

        let concurrency = state.active_uploads as f64 / self.config.concurrency_limit as f64;
        if concurrency >= self.config.token_warn_threshold {
            warn!(
                "Concurrent uploads at {}/{} ({:.0}% of limit)",
                state.active_uploads,
                self.config.concurrency_limit,
                concurrency * 100.0
            );
        }

        Ok(UploadGuard {
            monitor: self.clone(),
            size_bytes,
        })
    }

    fn finish_upload(&self, size_bytes: u64) {
        let mut state = self.state.lock().unwrap();
        state.active_uploads = state.active_uploads.saturating_sub(1);
        state.uploaded_bytes = state.uploaded_bytes.saturating_sub(size_bytes);
    }
}

pub struct UploadGuard {
    monitor: QuotaMonitor,
    size_bytes: u64,
}

impl Drop for UploadGuard {
    fn drop(&mut self) {
        self.monitor.finish_upload(self.size_bytes);
    }
}
